### PR TITLE
Upgrade to Keycloak 2.5.0

### LINKED ISSUE #47: Upgrade Keycloak to 2.5.0
Keycloak 2.5.0.Final was just released.  The release notes are up at http://blog.keycloak.org/2016/12/keycloak-250cr1-released.html.  I've run the `image-diff.sh` tool, diffing the 2.4.0.Final and 2.5.0.Final images, and there's only a couple files we'll likely need to tweak, both login-related templates:

```diff
diff -ru /tmp/keycloak/jboss_keycloak-postgres-2.4.0.Final/jboss/keycloak/themes/base/login/login-reset-password.ftl /tmp/keycloak/jboss_keycloak-postgres-2.5.0.Final/jboss/keycloak/themes/base/login/login-reset-password.ftl
--- /tmp/keycloak/jboss_keycloak-postgres-2.4.0.Final/jboss/keycloak/themes/base/login/login-reset-password.ftl	2016-11-23 15:37:56.000000000 -0500
+++ /tmp/keycloak/jboss_keycloak-postgres-2.5.0.Final/jboss/keycloak/themes/base/login/login-reset-password.ftl	2017-01-03 09:29:30.000000000 -0500
@@ -8,7 +8,7 @@
         <form id="kc-reset-password-form" class="${properties.kcFormClass!}" action="${url.loginAction}" method="post">
             <div class="${properties.kcFormGroupClass!}">
                 <div class="${properties.kcLabelWrapperClass!}">
-                    <label for="username" class="${properties.kcLabelClass!}"><#if !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if></label>
+                    <label for="username" class="${properties.kcLabelClass!}"><#if !realm.loginWithEmailAllowed>${msg("username")}<#elseif !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if></label>
                 </div>
                 <div class="${properties.kcInputWrapperClass!}">
                     <input type="text" id="username" name="username" class="${properties.kcInputClass!}" autofocus/>
diff -ru /tmp/keycloak/jboss_keycloak-postgres-2.4.0.Final/jboss/keycloak/themes/base/login/login.ftl /tmp/keycloak/jboss_keycloak-postgres-2.5.0.Final/jboss/keycloak/themes/base/login/login.ftl
--- /tmp/keycloak/jboss_keycloak-postgres-2.4.0.Final/jboss/keycloak/themes/base/login/login.ftl	2016-11-23 15:37:56.000000000 -0500
+++ /tmp/keycloak/jboss_keycloak-postgres-2.5.0.Final/jboss/keycloak/themes/base/login/login.ftl	2017-01-03 09:29:30.000000000 -0500
@@ -9,14 +9,14 @@
             <form id="kc-form-login" class="${properties.kcFormClass!}" action="${url.loginAction}" method="post">
                 <div class="${properties.kcFormGroupClass!}">
                     <div class="${properties.kcLabelWrapperClass!}">
-                        <label for="username" class="${properties.kcLabelClass!}"><#if !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if></label>
+                        <label for="username" class="${properties.kcLabelClass!}"><#if !realm.loginWithEmailAllowed>${msg("username")}<#elseif !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if></label>
                     </div>
 
                     <div class="${properties.kcInputWrapperClass!}">
                         <#if usernameEditDisabled??>
                             <input id="username" class="${properties.kcInputClass!}" name="username" value="${(login.username!'')?html}" type="text" disabled />
                         <#else>
-                            <input id="username" class="${properties.kcInputClass!}" name="username" value="${(login.username!'')?html}" type="text" autofocus />
+                            <input id="username" class="${properties.kcInputClass!}" name="username" value="${(login.username!'')?html}" type="text" autofocus autocomplete="off" />
                         </#if>
                     </div>
                 </div>
```

I'll hold off on this one until PRs #45 and #46 are wrapped up.

### PR DESCRIPTION
**Pull request by hkeeler**
Closes #47

I've upgrade the Docker image to `2.5.0.Final`.  There were a couple templates that had to be tweaked, and a cache-related change to `standalone.xml`, but that seems to be it!  I've done some local testing, and all appears to be working.

The one wrinkle is that they haven't pushed up a `2.5.0.Final` version of the [Adapters SPI](https://mvnrepository.com/artifact/org.keycloak/keycloak-adapter-spi) Java libs...yet?

I am currently building with `2.4.0.Final`, and again, it _seems_ fine, but I'd like to do a bit more testing before giving the green light to merge.

### CODE REVIEW

File: keycloak/Dockerfile [hunk 1]
```diff
@@ -1,11 +1,12 @@
-FROM jboss/keycloak-postgres:2.4.0.Final
+FROM jboss/keycloak-postgres:2.5.0.Final
 
 ENV MAVEN_VER=3.3.9
 ENV MAVEN_DIST=apache-maven-$MAVEN_VER
 ENV MAVEN_FILE=$MAVEN_DIST-bin.tar.gz
 ENV PATH=$PATH:/tmp/$MAVEN_DIST/bin
 ENV KC_SPI_SRC=providers
 ENV KC_SPI_DEST=/usr/src/keycloak
+ENV KC_VER=2.4.0.Final
```
Comment (wpears) [inline keycloak/Dockerfile]: What's the story with this vs. 2.5.0. on line 1?

Comment (hkeeler) [inline keycloak/Dockerfile]: This is the issue I mentioned on https://github.com/cfpb/hmda-platform-auth/pull/49#issue-198604487.  For whatever reason, they never published a `2.5.x` of the Keycloak Java libs.  `2.4.0.Final` libs _seem_ to work fine.

Perhaps I should put in a comment about that, and change the variable name to `KC_LIB_VER`.

File: keycloak/themes/hmda/login/login.ftl [hunk 2]
```diff
@@ -11,11 +11,11 @@
                     <legend class="usa-drop_text">Sign in</legend>
                     <span>or <a href="${url.registrationUrl}">create an account</a></span>
 
-                    <label for="username"><#if !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if></label>
+                    <label for="username"><#if !realm.loginWithEmailAllowed>${msg("username")}<#elseif !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if></label>
                     <#if usernameEditDisabled??>
                         <input id="username" name="username" type="text" autocapitalize="off" autocorrect="off" value="${(login.username!'')?html}" disabled>
                     <#else>
-                        <input id="username" name="username" value="${(login.username!'')?html}" type="text" autofocus />
+                        <input id="username" name="username" value="${(login.username!'')?html}" type="text" autofocus autocomplete="off" />
```
Comment (awolfe76) [inline keycloak/themes/hmda/login/login.ftl]: Did it happen to say why `autocomplete="off"`? Seems odd, not sure we need to keep it.

Comment (hkeeler) [inline keycloak/themes/hmda/login/login.ftl]: It's a bit strange.  It comes from https://github.com/keycloak/keycloak/commit/5c2bf7c822ec6a8fe72feb52872c452c34df100d, but the corresponding Github PR has no description, and the corresponding JIRA issue doesn't seem to be available to the public.

I can understand why they would make this change though.  Many sites don't allow username fields to be cached.

**Comment by awolfe76** [comment]
@hkeeler - once you add the comments (https://github.com/cfpb/hmda-platform-auth/pull/49#discussion_r94995322) we'll merge.

**Comment by hkeeler** [comment]
I updated the envvar in `keycloak/Dockerfile` for clarity, and added a warning comment to `login/theme.properties`.  I did a full _rebuild the world_ with the latest from all repos, and it all works great.  I think we're clear for merge.

**Comment by awolfe76** [comment]
Let me do the same "rebuild the world" and then I'll merge.

**Comment by awolfe76** [comment]
Looks good!

**Pull request merged**